feat: Add FactoryBot instrumentation

americodls opened the pull request:
Adds OpenTelemetry instrumentation for the FactoryBot gem to provide visibility into test data creation patterns during testing.

Subscribes to the factory_bot.run_factory ActiveSupport::Notifications event to capture all factory operations without monkey-patching. Supports FactoryBot 4.0+ and captures all factory strategies (create, build, build_stubbed, attributes_for) including batch operations (create_list, build_list, etc.).

Span naming: FactoryBot.{strategy}({factory_name})
Example: FactoryBot.build(user), FactoryBot.create_list(post)

Span attributes:
- factory_bot.strategy: Internal strategy name (build, create, stub, attributes_for)
- factory_bot.factory_name: Name of the factory being used
- factory_bot.traits: Comma-separated list of traits applied (if any)

americodls committed:
feat: Add FactoryBot instrumentation

Adds OpenTelemetry instrumentation for the FactoryBot gem to provide
visibility into test data creation patterns during testing.

Subscribes to the factory_bot.run_factory ActiveSupport::Notifications event
to capture all factory operations without monkey-patching. Supports FactoryBot
4.0+ and captures all factory strategies (create, build, build_stubbed,
attributes_for) including batch operations (create_list, build_list, etc.).

Span naming: FactoryBot.{strategy}({factory_name})
Example: FactoryBot.build(user), FactoryBot.create_list(post)

Span attributes:
- factory_bot.strategy: Internal strategy name (build, create, stub, attributes_for)
- factory_bot.factory_name: Name of the factory being used
- factory_bot.traits: Comma-separated list of traits applied (if any)

kaylareopelle commented:
Hi @americodls, thanks for opening this PR. Would you be interested in being a codeowner for this gem? That would involve reviewing PRs and issues related to the factory bot instrumentation. We don't have the codeowner behavior automated in GitHub, but we're currently documenting this in the CODEOWNERS file in the repo.

simi commented:
@americodls would you mind to share more how do you instrument testing also?

americodls commented on instrumentation/factory_bot/lib/opentelemetry/instrumentation/factory_bot/run_factory_subscriber.rb:
```diff
@@ -0,0 +1,61 @@
+          span_name = "FactoryBot.#{strategy_symbol}(#{factory_name})"
+
+          attrs = {
+            'factory_bot.strategy' => internal_strategy,
```
Thanks for the feedback!

I checked the [OpenTelemetry test semantic conventions](https://opentelemetry.io/docs/specs/semconv/attributes-registry/test/) and found that the test.* namespace currently only defines attributes for test execution metadata (test cases, suites, and their statuses).

Since test.data isn’t part of the spec yet, using it now could risk conflicts if it’s added later. The current factory_bot.* approach keeps things scoped to this library’s specific domain.

I’d suggest keeping the dedicated namespace for now — we can always generalize to something like test.data.* if similar fixture/factory patterns start appearing across other testing libraries. That way, we’re being intentional rather than speculative about the abstraction.

Of course, I’m open to other perspectives — input from the maintainers would be especially valuable given their broader experience in this space.

thompson-tomo replied:
So perhaps the following helps [defining attributes](https://opentelemetry.io/docs/specs/semconv/how-to-write-conventions/#defining-attributes)

> When defining an attribute for a narrow use case, think about potential broader use cases. For example, if creating a system-specific attribute, evaluate whether other systems in the same domain might need a similar attribute in the future.

I see the attributes identified as falling into that category as anything generating data needs a strategy etc.

I would even suggest raising a pr to get the attributes/spans added.

americodls replied:
> @americodls Are there any other testing libraries that are currently instrumented in other languages?
> 
> Is there enough traction there to start up a conversation between SIGs?

I haven't found anything that's why I suggested to "go as is".

arielvalentin reviewed (changes requested):
Thank you for your submission! 

There are a few required changes you will have to make to avoid issues/conflicts with our release tooling. 

I will have to reach out for help to other users who are FactoryBot users to help with a review before we merge.

americodls committed:
Update instrumentation/factory_bot/lib/opentelemetry/instrumentation/factory_bot/version.rb

Co-authored-by: Ariel Valentin <arielvalentin@users.noreply.github.com>

americodls committed:
Update instrumentation/factory_bot/opentelemetry-instrumentation-factory_bot.gemspec

Co-authored-by: Ariel Valentin <arielvalentin@users.noreply.github.com>

americodls committed:
Update instrumentation/factory_bot/CHANGELOG.md

Co-authored-by: Ariel Valentin <arielvalentin@users.noreply.github.com>

americodls committed:
fix: Change factory_bot.traits to array type

Per OpenTelemetry semantic conventions, plural attribute names should
use array values. Changed factory_bot.traits from comma-separated
string to string array.

americodls committed:
refactor: Improve test structure and coverage

- Move instrumentation.install to top-level before block
- Remove redundant install calls from nested describe blocks
- Add comprehensive attribute tests for all strategies
- Test empty traits array in all relevant sections
- Ensure all 3 attributes tested: strategy, factory_name, traits

americodls committed:
Merge branch 'main' into ad/add-factory-bot-instrumentation

americodls committed:
fix rubocop offense

americodls committed:
Merge remote-tracking branch 'upstream/main' into ad/add-factory-bot-instrumentation

americodls committed:
Hardcode ruby 3.2 as other libs do now

americodls committed:
Update instrumentation base version

americodls committed:
Merge branch 'main' into ad/add-factory-bot-instrumentation

arielvalentin reviewed (changes requested):
@americodls thank you again for your submission. 

Although I approved this PR, I am going to have request some additional changes. 

Our CI workflows have changed and this gems test suite is not longer running by default.

_You will need to add entries in the following actions workflow:_
- https://github.com/open-telemetry/opentelemetry-ruby-contrib/blob/1e018082a19f42a5f159838145b614d13537a5a4/.github/workflows/ci-instrumentation-full.yml#L8
- https://github.com/open-telemetry/opentelemetry-ruby-contrib/blob/1e018082a19f42a5f159838145b614d13537a5a4/.github/workflows/ci-instrumentation-full.yml#L67 

In addition to that please add `factorybot` as a gem dependency in our canary install test:

https://github.com/open-telemetry/opentelemetry-ruby-contrib/blob/1e018082a19f42a5f159838145b614d13537a5a4/releases/Gemfile#L14

Thank you for your patience.

americodls committed:
Add factory_bot to CI workflow and canary install test

americodls committed:
Merge branch 'main' into ad/add-factory-bot-instrumentation

americodls commented:
@arielvalentin

Thanks for your feedback!
I've gone ahead and fixed the issues you pointed out 🙂

Let me know if there’s anything else you’d like me to tweak!
